### ISSUE TITLE
potential leak of OVERLAPPED structures

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5372]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5372|
|Type|defect|
|Created|2011-11-07 19:18:51Z|
|Branch|https://github.com/twisted/twisted/tree/iocp-overlapped-5372|

When a socket function returns an error (other than ERROR_IO_PENDING), the iocpsupport module doesn't free the allocated OVERLAPPED structure before returning. If I'm not mistaken, this is a memory leak. Patch attached.

Attachments:

* [ovleaks.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9ca/9cab2d02817e2ac9d27c48c2a09edb901e5abf1e/872baeff763e8c9a82016ea7bdd2edbac12814b5.patch) (2722 bytes) - added by antoine on 2011-11-07 19:19:02Z - 
* [ovleaks2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9ca/9cab2d02817e2ac9d27c48c2a09edb901e5abf1e/fa18e950793b3ed8d44eb9ee9998c0e8e90f5d43.patch) (87294 bytes) - added by antoine on 2012-01-15 17:51:13Z - Updated patch with iocpsupport.c changes.

<details><summary>Searchable metadata</summary>

```
trac-id__5372 5372
type__defect defect
reporter__antoine antoine
priority__highest highest
milestone__ 
branch__branches_iocp_overlapped_5372 branches/iocp-overlapped-5372
branch_author__antoine antoine
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1320693531000000 1320693531000000
changetime__1331594237000000 1331594237000000
version__None None
owner__jesstess jesstess
cc__antoine cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess commented|
|-|-|

Thanks for the bug report and patch, antoine, and sorry for the delay in getting this reviewed. I think I've convinced myself via the msdn docs that your changes look good, but I'm not seeing where the OVERLAPPED structures are getting freed in the non-error case. Can you shed some light on that?

I'll stick this patch in a branch to kick off test builds, but for ease of testing can you attach the generated iocpsupport.c changes as well?

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_4"></a>|@pitrou commented|
|-|-|

Replying to [jesstess](#note_3):
> Thanks for the bug report and patch, antoine, and sorry for the delay in getting this reviewed. I think I've convinced myself via the msdn docs that your changes look good, but I'm not seeing where the OVERLAPPED structures are getting freed in the non-error case. Can you shed some light on that?

When the overlapped operation is initiated successfully (i.e. returns without error or with ERROR_IO_PENDING, the system libraries have ownership of the OVERLAPPED structure. Completion is then checked using GetQueuedCompletionStatus() (in CompletionPort.getEvent in iocpsupport.pyx) which gives back the OVERLAPPED structure which has completed. This is where we finally free the structure.

(note I'm not a Win32 expert, but this is what I get following the MSDN docs)

I'm attaching a patch including iocpsupport.c changes.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess commented|
|-|-|

(In [33436]) Branching to 'iocp-overlapped-5372'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess commented|
|-|-|

(In [33437]) Apply patch ovleaks2.patch by antoine

This includes the generate iocpsupport.c changes.

refs [#5372](https://github.com/twisted/twisted/issues/5372)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|PenguinOfDoom set owner to @jesstess|
|-|-|

I approve.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set status to `closed`|
|-|-|

Merge iocp-overlapped-5372

Author: antoine
Reviewer: jesstess, PenguinOfDoom
Fixes: [#5372](https://github.com/twisted/twisted/issues/5372)

Fix a leak in some error cases of OVERLAPPED structures.
